Fire an event when failing to open NodePort

masap opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
[issue]
When creating a NodePort service with the ```kubectl create``` command, the NodePort assignment may fail.
Failure to assign a NodePort can be simulated with the following malicious command[1].
```
$ kubectl create service nodeport temp-svc --tcp=`python3 <<EOF
print("1", end="")
for i in range(2, 1026):
  print("," + str(i), end="")
EOF
`
```
The command succeeds and shows following output.
```
service/temp-svc created
```
The service has been successfully generated and can also be referenced with the get command.
```
$ kubectl get svc
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)
temp-svc     NodePort    10.0.0.139   <none>        1:31335/TCP,2:32367/TCP,3:30263/TCP,(omitted),1023:31821/TCP,1024:32475/TCP,1025:30311/TCP   12s
```
The user does not recognize failure to assign a NodePort because create/get/describe command does not show any error. This is the issue.

[solution]
Users can notice errors by looking at the ```kube-proxy``` logs, but it may be difficult to see the ```kube-proxy``` logs of all nodes.
```
E0327 08:50:10.216571  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :30641: socket: too many open files" port="\"nodePort for default/temp-svc:744\" (:30641/tcp4)"
E0327 08:50:10.216611  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :30827: socket: too many open files" port="\"nodePort for default/temp-svc:857\" (:30827/tcp4)"
...
E0327 08:50:10.217119  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :32484: socket: too many open files" port="\"nodePort for default/temp-svc:805\" (:32484/tcp4)"
E0327 08:50:10.217293  660960 proxier.go:1612] "Failed to execute iptables-restore" err="pipe2: too many open files ()"
I0327 08:50:10.217341  660960 proxier.go:1615] "Closing local ports after iptables-restore failure"
```
So, this patch will fire an event when NodePort assignment fails.
In fact, when the externalIP assignment fails, it is also notified by event.

The event will be displayed like this.
```
$ kubectl get event
LAST SEEN   TYPE      REASON                                            OBJECT           MESSAGE
...
2s          Warning   listen tcp4 :31055: socket: too many open files   node/127.0.0.1   can't open "nodePort for default/temp-svc:901" (:31055/tcp4), skipping this nodePort: listen tcp4 :31055: socket: too many open files
2s          Warning   listen tcp4 :31422: socket: too many open files   node/127.0.0.1   can't open "nodePort for default/temp-svc:474" (:31422/tcp4), skipping this nodePort: listen tcp4 :31422: socket: too many open files
...
```

This PR fixes ```iptables``` and ```ipvs``` proxier. Since ```userspace``` proxier does not seem to be affected by this issue, it is not fixed.

[1] Assume that fd limit is 1024(default).
```
$ ulimit -n
1024
```
#### Which issue(s) this PR fixes:
NONE

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fire an event when failing to open NodePort
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

masap committed:
Fire an event when failing to open NodePort

[issue]
When creating a NodePort service with the kubectl create command, the NodePort
assignment may fail.

Failure to assign a NodePort can be simulated with the following malicious
command[1].

$ kubectl create service nodeport temp-svc --tcp=`python3 <<EOF
print("1", end="")
for i in range(2, 1026):
  print("," + str(i), end="")
EOF
`

The command succeeds and shows following output.

service/temp-svc created

The service has been successfully generated and can also be referenced with the
get command.

$ kubectl get svc
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)
temp-svc     NodePort    10.0.0.139   <none>        1:31335/TCP,2:32367/TCP,3:30263/TCP,(omitted),1023:31821/TCP,1024:32475/TCP,1025:30311/TCP   12s

The user does not recognize failure to assign a NodePort because
create/get/describe command does not show any error. This is the issue.

[solution]
Users can notice errors by looking at the kube-proxy logs, but it may be difficult to see the kube-proxy logs of all nodes.

E0327 08:50:10.216571  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :30641: socket: too many open files" port="\"nodePort for default/temp-svc:744\" (:30641/tcp4)"
E0327 08:50:10.216611  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :30827: socket: too many open files" port="\"nodePort for default/temp-svc:857\" (:30827/tcp4)"
...
E0327 08:50:10.217119  660960 proxier.go:1286] "can't open port, skipping this nodePort" err="listen tcp4 :32484: socket: too many open files" port="\"nodePort for default/temp-svc:805\" (:32484/tcp4)"
E0327 08:50:10.217293  660960 proxier.go:1612] "Failed to execute iptables-restore" err="pipe2: too many open files ()"
I0327 08:50:10.217341  660960 proxier.go:1615] "Closing local ports after iptables-restore failure"

So, this patch will fire an event when NodePort assignment fails.
In fact, when the externalIP assignment fails, it is also notified by event.

The event will be displayed like this.

$ kubectl get event
LAST SEEN   TYPE      REASON                                            OBJECT           MESSAGE
...
2s          Warning   listen tcp4 :31055: socket: too many open files   node/127.0.0.1   can't open "nodePort for default/temp-svc:901" (:31055/tcp4), skipping this nodePort: listen tcp4 :31055: socket: too many open files
2s          Warning   listen tcp4 :31422: socket: too many open files   node/127.0.0.1   can't open "nodePort for default/temp-svc:474" (:31422/tcp4), skipping this nodePort: listen tcp4 :31422: socket: too many open files
...

This PR fixes iptables and ipvs proxier.
Since userspace proxier does not seem to be affected by this issue, it is not fixed.

[1] Assume that fd limit is 1024(default).
$ ulimit -n
1024

masap commented:
/test pull-kubernetes-unit

aojea commented:
heh, we just talked about this recently
/assign @thockin

aojea commented:
/test pull-kubernetes-e2e-kind-dual-canary

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1283,6 +1283,15 @@ func (proxier *Proxier) syncProxyRules() {
 				} else if svcInfo.Protocol() != v1.ProtocolSCTP {
 					socket, err := proxier.portMapper.OpenLocalPort(&lp)
 					if err != nil {
+						msg := fmt.Sprintf("can't open %s, skipping this nodePort: %v", lp.String(), err)
```
How about "can't open NodePort %s, skipping it: %v" - simpler

masap replied:
Ok. I fixed it by additional commit to fix similar other messages.

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1466,6 +1466,15 @@ func (proxier *Proxier) syncProxyRules() {
 				} else if svcInfo.Protocol() != v1.ProtocolSCTP {
 					socket, err := proxier.portMapper.OpenLocalPort(&lp)
 					if err != nil {
+						msg := fmt.Sprintf("can't open %s, skipping this nodePort: %v", lp.String(), err)
```
same

masap replied:
Fixed as well.

thockin reviewed (commented):
I am OK to merge this for now, but I do want to re-open the discussion about NOT doing this.  @aojea - did we write that down?

masap commented:
/test pull-kubernetes-e2e-kind-dual-canary

masap commented:
/test pull-kubernetes-e2e-kind

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1110,7 +1110,7 @@ func (proxier *Proxier) syncProxyRules() {
 					socket, err := proxier.portMapper.OpenLocalPort(&lp)
 					if err != nil {
-						msg := fmt.Sprintf("can't open %s, skipping this externalIP: %v", lp.String(), err)
+						msg := fmt.Sprintf("can't open ExternalIP %s, skipping it: %v", lp.String(), err)
```
Sorry, this still isn't quite right - the lp.String() will say "externalIP" so simply `"can't open port: %s: %v", lp.String(), err` should suffice in all of these

masap replied:
OK, fixed. Now the messages are showed like this.
```
$ kubectl get event
5m52s       Warning   listen tcp4 :32179: socket: too many open files   node/127.0.0.1   can't open port "nodePort for default/temp-svc:370" (:32179/tcp4), skipping it: listen tcp4 :32179: socket: too many open files
```

```kube-proxy``` log
```
E0331 09:42:08.238370   40517 proxier.go:1295] "can't open port, skipping it" err="listen tcp4 :32179: socket: too many open files" port="\"nodePort for default/temp-svc:370\" (:32179/tcp4)"
```


As for ```userspace```, it is left as it is, because ```key.String()``` does not contain a string describing port type. Like "10.0.0.1:443/TCP". The messages in ```kube-proxy``` log is like this.
```
E0331 10:20:44.586170   93970 proxier.go:415] "Failed to ensure portal" err="can't open node port for 10.0.0.1:443/TCP: <nil>" servicePortName="default/kubernetes:https"
```

masap commented:
/test pull-kubernetes-e2e-kind-dual-canary

masap commented:
/test pull-kubernetes-e2e-kind

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1110,7 +1110,7 @@ func (proxier *Proxier) syncProxyRules() {
 					socket, err := proxier.portMapper.OpenLocalPort(&lp)
 					if err != nil {
-						msg := fmt.Sprintf("can't open %s, skipping this externalIP: %v", lp.String(), err)
+						msg := fmt.Sprintf("can't open port %s, skipping it: %v", lp.String(), err)
```
I am so sorry to pick on this so hard.  Since you posted the output (THANK YOU) I see the error is reported twice.  I think we do not need the `%v` and `err` in msg.  That makes it about the same as the log-line

masap replied:
OK, OK, fixed.
```
$ kubectl get event
LAST SEEN   TYPE      REASON                                            OBJECT           MESSAGE
4s          Warning   listen tcp4 :30486: socket: too many open files   node/127.0.0.1   can't open port "nodePort for default/temp-svc:503" (:30486/tcp4), skipping it
```

masap committed:
Use simpler expressions for error messages

1. Do not describe port type in message because lp.String() already has the
information.

2. Remove duplicate error detail from event log.
Previous log is like this.

47s         Warning   listen tcp4 :30764: socket: too many open files   node/127.0.0.1   can't open port "nodePort for default/temp-svc:834" (:30764/tcp4), skipping it: listen tcp4 :30764: socket: too many open files

masap commented:
/test pull-kubernetes-e2e-kind-dual-canary

masap commented:
/test pull-kubernetes-e2e-kind-dual-canary

thockin commented:
Thanks!

/lgtm
/approve
